text/0000-async-stream.md

```diff
@@ -0,0 +1,758 @@
+    stream: &'a mut S,
+}
+
+impl<St: ?Sized + Unpin> Unpin for Next<'_, St> {}
+
+impl<'a, St: ?Sized + Stream + Unpin> Next<'a, St> {
+    pub(super) fn new(stream: &'a mut St) -> Self {
+        Next { stream }
+    }
+}
+
+impl<St: ?Sized + Stream + Unpin> Future for Next<'_, St> {
+    type Output = Option<St::Item>;
+
+    fn poll(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Self::Output> {
+        Pin::new(&mut *self.stream).poll_next(cx)
+    }
+}
+```
+
+This would allow a user to await on a future:
+
+```rust
+while let Some(v) = stream.next().await {
+
+}
+```
+
+We could also consider adding a try_next? function, allowing
+a user to write:
+
+```rust
+while let Some(x) = s.try_next().await?
+```
+
+One thing to note, if a user is using an older version of `futures-util`,
+they would experience ambiguity when trying to use the `next` method that
+is added to the standard library (and redirected to from `futures-core`).
+
+This can be done as a non-breaking change, but would require everyone to 
+upgrade rustc. We will want to create a transition plan on what this
+means for users and pick the timing carefully.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This section goes into details about various aspects of the design and
+why they ended up the way they did.
+
+## Where does `Stream` live in the std lib?
+
+`Stream` will live in the `core::stream` module and be re-exported as `std::stream`.
+
+It is possible that it could live in another area as well, though this followes
+the pattern of `core::future`.
+
+## Why use a `poll` method?
+
+An alternative design for the stream trait would be to have a trait
+that defines an async `next` method:
+
+```rust
+trait Stream {
+    type Item;
+    
+    async fn next(&mut self) -> Option<Self::Item>;
+}
+```
+
+Unfortunately, async methods in traits are not currently supported,
+and there [are a number of challenges to be
+resolved](https://rust-lang.github.io/wg-async-foundations/design_notes/async_fn_in_traits.html)
+before they can be added. 
+
+Moreover, it is not clear yet how to make traits that contain async
+functions be `dyn` safe, and it is imporant to be able to pass around `dyn
+Stream` values without the need to monomorphize the functions that work
+with them.
+
+Unfortunately, the use of poll does mean that it is harder to write
+stream implementations. The long-term fix for this, discussed in the [Future possiblilities](future-possibilities) section, is dedicated [generator syntax].
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Where should stream live?
+
+As mentioned above, `core::stream` is analogous to `core::future`. But, do we want to find 
+some other naming scheme that can scale up to other future additions, such as io traits or channels?
```

Comment:
maybe `core::future::Stream`?

because it's clearer to say that `std::net::TcpStream` is not a `future::Stream` than a `stream::Stream` 🙃

Reply:
I would be open to this :)

Reply:
Putting `Future` and `Stream` in the same hierarchy could be problematic since for example `async_std` exposes both `future::{pending, Pending}` and `stream::{pending, Pending}`. I expect the stdlib to want to do the same, which would not be possible under this proposal.

An alternate option would be to expose `Stream` under `core::future::stream`, but personally I'm not a fan of nesting it this way. To my knowledge the only other location in std where that's done is in the `std::sync::{atomic, mpsc}` submodules, and the libs team is currently discussing how to break that up to be flatter and more tightly scoped.